ownership/sync-repos.py

```diff
@@ -0,0 +1,255 @@
+                {'new_location': str(found_repo)})
+        elif found_repo.archived:
+            recommend_delete_row(ss_name, "Repo has been archived")
+        else:
+            new_name = str(found_repo).lower()
+            if [r for r in repos_ss if r['name'] == new_name]:
+                bail("Manual intervention required: A repo that has moved "
+                     "exists in the spreadsheet under both its old name "
+                     "and its new name. %s -> %s"
+                     % (ss_name, new_name))
+            actions.append({
+                'action': 'rename_row',
+                'name': ss_name,
+                'new_name': new_name,
+                'row_id': find_row_by_name(ss_name)['row_id'],
+                'why': ("Repo has been renamed, and is not already in the "
+                        "spreadsheet under that name")
+            })
+            names_only_ss.remove(ss_name)
+            names_only_gh.remove(new_name)
+    
+    # Now check ones that are only in Github. We've already filtered
+    # out archived repos and removed any that are pending a rename on
+    # the spreadsheet.
+    for gh_name in names_only_gh.copy():
+        actions.append({
+            'action': 'add_row',
+            'name': gh_name,
+            'why': "Repo present in Github but not spreadsheet"
+        })
+        names_only_gh.remove(gh_name)
+
+    # Make sure nothing is unaccounted for before presenting recommendations.
+    if names_only_gh or names_only_ss:
+        bail("Failed to convert all mismatches between Github and spreadsheet "
+             "into actions! Github remaining = %s; spreadsheet remaining = %s"
+             % (names_only_gh, names_only_ss))
+
+    # Ensure actions don't have any internal conflicts. It might be OK
+    # if the same *name* comes up twice; for example, we might want to
+    # move an archived repo to the unspported org *and* remove it from
+    # the spreadsheet. However, two actions for the same row ID is a
+    # problem.
+    seen_row_ids = set()
+    for action in actions:
+        row_id = action.get('row_id')
+        if not row_id:
+            continue
+        if row_id in seen_row_ids:
+            bail("Possible bug in script: Was going to suggest multiple "
+                 "actions for row %s: %s"
+                 % (row_id, [r for r in actions if r.get('row_id') == row_id]))
+        seen_row_ids.add(row_id)
+
+    # For now, just make recommendations, don't make changes.
+    if actions:
+        print("Recommendations:")
+
+        groupkey = lambda a: (a['action'], a['why'])
+        for ((action_name, why), group) in groupby(sorted(actions, key=groupkey), groupkey):
+            print()
+            print("Action: %s -- %s" % (action_name, why))
+            for action in group:
+                data = action.copy()
+                del data['action']
+                del data['name']
+                del data['why']
```

Comment:
There has to be a better way to disassociate multiple keys from a copy of dictionary, right? In Clojure this would be `(dissoc data :action :name :why)` but I can't figure out a nice equivalent for Python.

Reply:
Why does this need to be done?

Reply:
There's extra data associated with some action types, but not all. This removes the keys that are always present, since they're printed explicitly, and then `data` can be listed out if there's anything left. Sample output:

```
Action: delete_row -- Moved to a Github org not in the scan list
- edx/cookiecutter-django-app {'row_id': 26, 'new_location': 'edx-unsupported/cookiecutter-django-app'}
- edx/edx-oauth2-provider {'row_id': 25, 'new_location': 'edx-unsupported/edx-oauth2-provider'}

Action: delete_row -- Repo has been archived
- edx/eslint-config-edx {'row_id': 244}

Action: move_repo_unsupported -- Repo has been archived but is still in a main org rather than edx-unsupported
- edx/edx-ora
```

If each action type were handled by a different code branch, I would pull out relevant attributes explicitly for display, but this was a simpler way. This is probably worth refining if called for, since I bet we'll want a `--dry-run` flag even once this can make changes.

Incidentally, I did figure out a way of doing this as a one-liner, but I'm not sure if it would make people sad:

```python
data = {k: v for k, v in action.items() if k not in {'action', 'name', 'why'}}
```